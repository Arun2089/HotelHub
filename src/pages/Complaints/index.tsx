import { useState, MouseEvent, useContext, ChangeEvent } from "react";
import {
  Box,
  Typography,
  DialogContent,
  MenuItem,
  Select,
  SelectChangeEvent,
  IconButton,
  Button,
} from "@mui/material";
import {
  GridColDef,
  GridColumnHeaderParams,
  GridAlignment,
} from "@mui/x-data-grid";
import { DeleteOutlineOutlined } from "@mui/icons-material";

import TableComponent from "@components/Table";
import AlertComponent from "@components/Alert";
import ConfirmationModal from "@components/ConfirmationModal";
import ChipComponent from "@src/components/Clip";
import ComplaintModal from "./ComplaintModal";

import useDialog from "@src/hooks/useDialog";
import useAlert from "@src/hooks/useAlert";
import { AuthContext } from "@context/AuthContext";
import {
  getStatusColor,
  dateFormat,
  sendData,
  fetchData,
  catchErrorMessage,
  extractArrayFromApiData,
  todayDate,
  deleteData,
} from "@utils/index";
import {
  COMPLAINTS_URL,
  COMPLAINT_STATUS,
  ERROR,
  STUDENT,
  SUCCESS,
  STATUS_ICONS,
} from "@src/constant";
import {
  ComplaintType,
  SeverityType,
  ComplaintStateType,
  ComplaintStatusType,
} from "@ts/types";

type SearchStatusType = "all" | "pending" | "resolved" | "invalid";

const updateComplaintsStatus = async (
  complaint: ComplaintType,
  handleAlert: (
    isOpen: boolean,
    message: string,
    serverity: SeverityType
  ) => void
) => {
  const { id } = complaint;
  const url = `${COMPLAINTS_URL}/${id}`;
  try {
    const isDataUpdated = await sendData<ComplaintType>(url, "PUT", complaint);
    const message = isDataUpdated
      ? "Status saved Successfully"
      : "Error in updating status";
    const severity = isDataUpdated ? SUCCESS : ERROR;
    handleAlert(true, message, severity);
  } catch (error) {
    handleAlert(true, catchErrorMessage(error), ERROR);
  }
};

const deleteComplaint = async (
  { id }: ComplaintType,
  handleAlert: (
    isOpen: boolean,
    message: string,
    serverity: SeverityType
  ) => void
) => {
  const url = `${COMPLAINTS_URL}/${id}`;
  try {
    const isDataDeleted = await deleteData(url);
    const message = isDataDeleted
      ? "Complaint Deleted Successfully"
      : "Error in deleting complaint";
    const severity = isDataDeleted ? SUCCESS : ERROR;
    handleAlert(true, message, severity);
  } catch (error) {
    handleAlert(true, catchErrorMessage(error), ERROR);
  }
};

const isMenuItemVisible = (optionStatus: string, status: string) => {
  if (optionStatus === "pending") return true;
  if (optionStatus === "resolved" && status !== "pending") return true;
  if (optionStatus === "invalid" && status !== "pending") return true;
  return false;
};

const commonOptions: {
  sortable: boolean;
  align: GridAlignment;
  headerAlign: GridAlignment;
  renderHeader: (params: GridColumnHeaderParams) => JSX.Element;
} = {
  sortable: false,
  align: "center",
  headerAlign: "center",
  renderHeader({ colDef: { headerName } }: GridColumnHeaderParams) {
    return <Typography className="font-semibold">{headerName}</Typography>;
  },
};

const Complaints = () => {
  const [complaintData, setComplaintData] = useState<ComplaintType[] | null>(
    null
  );
  const [loading, setLoading] = useState<boolean>(false);
  const [selectedComplaint, setSelectedComplaint] =
    useState<ComplaintStateType | null>(null);
  const [searchStatus, setSearchStatus] = useState<SearchStatusType>("all");
  const [rowCount, setRowCount] = useState<number>(0);
  const [paginationModel, setPaginationModel] = useState({
    page: 0,
    pageSize: 10,
  });
  const [action, setAction] = useState<string>("");
  const {
    user: { role, name },
  } = useContext(AuthContext);
  const isRoleStudent = role === STUDENT;
  const { status: seletedComplaintStatus } = selectedComplaint?.complaint || {};
  const { open, handleDialogClick, handleDialogSubmit } = useDialog(
    async () => {
      action === "delete"
        ? await deleteComplaint(selectedComplaint!.complaint, handleAlert)
        : await updateComplaintsStatus(
            selectedComplaint!.complaint,
            handleAlert
          );
      getComplaintsData(true, paginationModel.page, searchStatus);
    }
  );
  const { alert, handleAlert } = useAlert();
  const selectMenuItems = [
    { status: "pending", text: "Pending", icon: STATUS_ICONS.pending },
    { status: "resolved", text: "Resolved", icon: STATUS_ICONS.resolved },
    { status: "invalid", text: "Invalid", icon: STATUS_ICONS.invalid },
  ];
  const initialComplaint: ComplaintType = {
    date: todayDate(),
    type: "",
    description: "",
    status: "pending",
    studentName: name,
  };
  async function getComplaintsData(
    pagination: boolean,
    page?: number,
    searchStatus?: string
  ) {
    try {
      setLoading(true);
      const response = await fetchData(COMPLAINTS_URL);
      const data = extractArrayFromApiData<ComplaintType>(response.data);
      data.sort(
        (a: ComplaintType, b: ComplaintType) =>
          new Date(b.date).getTime() - new Date(a.date).getTime()
      );
      const filteredData = searchStatus
        ? data.filter(({ status }: ComplaintType) => {
            if (searchStatus === "all") return true;
            return status === searchStatus;
          })
        : data;
      filteredData.forEach((complaint: ComplaintType) => {
        complaint.date = dateFormat(complaint.date);
      });

      if (!pagination) setComplaintData(filteredData);
      else {
        setComplaintData(filteredData.slice(page! * 10, page! * 10 + 10));
      }
      setRowCount(filteredData.length);
      setLoading(false);
    } catch (error) {
      setLoading(false);
      handleAlert(true, catchErrorMessage(error), "error");
    }
  }

  const addNewComplaint = async (
    complaint: ComplaintType,
    handleAlert: (
      isOpen: boolean,
      message: string,
      serverity: SeverityType
    ) => void
  ) => {
    try {
      const isComplaintAdded = await sendData<ComplaintType>(
        COMPLAINTS_URL,
        "POST",
        complaint
      );
      const message = isComplaintAdded
        ? "Complaint added Successfully"
        : "Error in adding complaint";
      const severity = isComplaintAdded ? SUCCESS : ERROR;
      handleAlert(true, message, severity);
      isComplaintAdded &&
        getComplaintsData(true, paginationModel.page, searchStatus);
      setSelectedComplaint(null);
    } catch (error) {
      handleAlert(true, catchErrorMessage(error), ERROR);
    }
  };

  const handleOpenModal = (
    event: MouseEvent<Element>,
    row: unknown,
    isModalEditable: boolean = false
  ) => {
    event.stopPropagation();
    setSelectedComplaint({
      complaint: row as ComplaintType,
      isModalOpen: true,
      isModalEditable,
    });
  };

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setSelectedComplaint({
      ...selectedComplaint!,
      complaint: {
        ...selectedComplaint!.complaint,
        [name]: value,
      },
    });
  };

  const handleAutoCompleteChange = (type: string | null) => {
    setSelectedComplaint({
      ...selectedComplaint!,
      complaint: {
        ...selectedComplaint!.complaint,
        type: type || "",
      },
    });
  };

  const OnStatusChange = (
    event: SelectChangeEvent<"pending" | "resolved" | "invalid">,
    row: ComplaintType
  ) => {
    event.stopPropagation();
    const { value } = event.target;
    setSelectedComplaint({
      complaint: { ...row, status: value as ComplaintStatusType },
      isModalOpen: false,
      isModalEditable: false,
    });
    setAction("status");
    handleDialogClick(true);
  };

  const adminSpecificColumns: GridColDef[] = [
    {
      field: "studentName",
      headerName: "Student",
      minWidth: 170,

      flex: 1,
      cellClassName: "hover:cursor-pointer",
      ...commonOptions,
    },
  ];

  const studentSpecificColumns: GridColDef[] = [
    {
      field: "status",
      headerName: "Status",
      minWidth: 170,
      cellClassName: "hover:cursor-pointer",
      ...commonOptions,
      renderCell: ({ row }) => {
        const { status } = row as ComplaintType;
        const { textColor } = getStatusColor(status, true);
        const Icon = STATUS_ICONS[status];
        return (
          <ChipComponent
            className="w-24 capitalize"
            text={status}
            type={status}
            icon={<Icon fontSize="small" className={textColor} />}
          />
        );
      },
    },
  ];

  const columns: GridColDef[] = [
    {
      field: "id",
      headerName: "Id",
      minWidth: 50,
      headerClassName: "pl-8",
      cellClassName: "pl-8 hover:cursor-pointer",
      ...commonOptions,
    },
    {
      field: "date",
      headerName: "Date",
      minWidth: 150,
      headerClassName: "pl-8",
      cellClassName: "pl-8 hover:cursor-pointer",
      ...commonOptions,
    },
    {
      field: "type",
      headerName: "Type",

      minWidth: 170,
      flex: 1,
      cellClassName: "hover:cursor-pointer",
      ...commonOptions,
    },
    {
      field: "description",
      headerName: "Description",
      minWidth: 170,

      flex: 1,
      cellClassName: "hover:cursor-pointer",
      ...commonOptions,
    },
    ...(isRoleStudent ? studentSpecificColumns : adminSpecificColumns),
    {
      field: "Actions",
      headerName:
        isRoleStudent || searchStatus === "pending" || searchStatus === "all"
          ? "Actions"
          : "Status",
      minWidth: 170,
      cellClassName: "hover:cursor-pointer",
      ...commonOptions,
      renderCell: ({ row }) => {
        const { status } = row as ComplaintType;
        const isPending = status === "pending";
        return (
          <>
            {isRoleStudent ? (
              <IconButton
                onClick={(event) => {
                  event.stopPropagation();
                  setSelectedComplaint((prev) => {
                    return {
                      ...prev!,
                      complaint: row as ComplaintType,
                      isModalOpen: false,
                      isModalEditable: false,
                    };
                  });
                  setAction("delete");
                  handleDialogClick(true);
                }}
                size="large"
                disabled={!isPending}
                className={`p-2 ${
                  isPending ? "text-error-main" : "text-gray-400"
                }`}
              >
                <DeleteOutlineOutlined />
              </IconButton>
            ) : (
              <Select
                sx={{
                  "& fieldset": { border: "none" },
                }}
                inputProps={{
                  className: "p-0",
                }}
                value={status}
                label="Age"
                IconComponent={() => null}
                onChange={(
                  event: SelectChangeEvent<"pending" | "resolved" | "invalid">
                ) => {
                  OnStatusChange(event, row as ComplaintType);
                }}
                MenuProps={{
                  PaperProps: {
                    style: {
                      marginTop: "0.5rem",
                    },
                  },
                }}
              >
                {selectMenuItems.map(
                  ({ status: optionStatus, icon: Icon, text }) => {
                    const { textColor } = getStatusColor(status, true);
                    const disabled = isMenuItemVisible(optionStatus, status);
                    return (
                      <MenuItem
                        key={optionStatus}
                        value={optionStatus}
                        disabled={disabled}
                      >
                        <ChipComponent
                          className="w-24"
                          text={text}
                          type={optionStatus}
                          icon={<Icon fontSize="small" className={textColor} />}
                        />
                      </MenuItem>
                    );
                  }
                )}
              </Select>
            )}
          </>
        );
      },
    },
  ];

  return (
    <>
      <Box className="flex w-[96%] justify-center md:justify-start gap-2 shrink flex-wrap mx-auto gap-x-4">
        {COMPLAINT_STATUS.map(({ status, icon: Icon }) => {
          const isActive = searchStatus === status;
          const { textColor } = getStatusColor(status, isActive);
          return (
            <ChipComponent
              key={status}
              text={status.toUpperCase()}
              type={status}
              className="w-28 justify-around"
              isActive={isActive}
              variant={isActive ? "filled" : "outlined"}
              icon={<Icon fontSize="small" className={textColor} />}
              onClick={() => {
                setSearchStatus(status as SearchStatusType);
                setPaginationModel({ page: 0, pageSize: 10 });
                getComplaintsData(true, paginationModel.page, status);
              }}
            />
          );
        })}
        {isRoleStudent && (
          <Box className="flex-1 flex">
            <Button
              variant="contained"
              className="ml-auto"
              size="large"
              onClick={(event) => {
                handleOpenModal(event, initialComplaint, true);
              }}
            >
              raise complaint
            </Button>
          </Box>
        )}
      </Box>
      <Box className="w-[96%] h-[85vh] md:h-full my-4 xl:my-6  mx-auto overflow-hidden rounded-xl">
        <TableComponent
          columns={columns}
          isLoading={loading}
          rows={complaintData || []}
          tableClassName="max-h-full bg-white rounded-xl"
          pagination={true}
          getData={getComplaintsData}
          rowCount={rowCount}
          paginationModel={paginationModel}
          setPaginationModel={setPaginationModel}
          onRowClick={(event: MouseEvent<Element>, row) => {
            handleOpenModal(event, row);
          }}
          searchText={searchStatus}
        />
        {alert.isOpen && (
          <AlertComponent
            severity={alert.severity}
            message={alert.message}
            handleClose={() => handleAlert(false)}
          />
        )}
        <ConfirmationModal
          isOpen={open}
          handleClose={() => handleDialogClick(false)}
          handleSubmit={handleDialogSubmit}
          title="Are You Sure ?"
          buttontext={action === "delete" ? "delete" : "save"}
          buttonType={action === "delete" ? "delete" : "primary"}
        >
          <DialogContent className="padding-0 ">
            {action === "delete"
              ? "Complaint will be deleted"
              : `Complaint Status will be changed to ${seletedComplaintStatus}`}
          </DialogContent>
        </ConfirmationModal>
        {selectedComplaint?.isModalOpen && (
          <ComplaintModal
            complaintState={selectedComplaint}
            handleClose={() => setSelectedComplaint(null)}
            handleChange={handleInputChange}
            handleAutoCompleteChange={handleAutoCompleteChange}
            handleAdd={() => {
              addNewComplaint(selectedComplaint!.complaint, handleAlert);
            }}
          />
        )}
      </Box>
    </>
  );
};

export default Complaints;
