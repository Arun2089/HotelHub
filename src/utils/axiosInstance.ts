import axios, { AxiosInstance } from "axios";
import { BASE_URL } from "@constant/index";
import { deleteLocalStorage, getLocalStorage } from "@utils/index";

const axiosInstance = (): AxiosInstance => {
  const instance = axios.create({
    baseURL: BASE_URL,
  });
  instance.interceptors.request.use(
    (config) => {
      const token = getLocalStorage("jwtToken");
      if (token) config.headers.Authorization = `Bearer ${token}`;
      else window.location.href = "/login";
      return config;
    },
    (error) => Promise.reject(error)
  );

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      if (error.response.status === 401) {
        deleteLocalStorage("jwtToken");
        window.location.href = "/login";
      }
      return Promise.reject(error);
    }
  );

  return instance;
};

export default axiosInstance;
