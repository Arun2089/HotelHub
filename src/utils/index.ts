import { StudentInfoType } from "@ts/types";
import { STUDENT_INFO_URL } from "@constant/index";

export const fetchData = async (url: string) => {
  const response = await fetch(url);
  if (response.ok) {
    const data = await response.json();
    return data;
  } else {
    throw new Error(`Error in retrieving data from ${url}`);
  }
};
export const sendData = async (
  url: string,
  method: "POST" | "PUT",
  data: any //eslint-disable-line
): Promise<boolean> => {
  const response = await fetch(url, {
    method: method,
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });
  if (response.ok) {
    return true;
  } else {
    throw new Error(`Error in sending data`);
  }
};

export const deleteData = async (url: string): Promise<boolean> => {
  const response = await fetch(url, {
    method: "DELETE",
  });
  if (response.ok) {
    return true;
  } else {
    throw new Error(`Error in deleting data from ${url}`);
  }
};

export const checkEmailExists = async (
  email: string,
  id?: number
): Promise<boolean> => {
  try {
    const data = await fetchData(STUDENT_INFO_URL);
    if (id) {
      const emailExists = data.some(
        (student: StudentInfoType) =>
          student.email === email && student.id !== id
      );
      return emailExists;
    }
    const emailExists = data.some(
      (student: StudentInfoType) => student.email === email
    );
    return emailExists;
  } catch (error) {
    throw error as string;
  }
};

export const catchErrorMessage = (error: unknown) => {
  return error instanceof Error ? error.message : "An error occurred";
};
