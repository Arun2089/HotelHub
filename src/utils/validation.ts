function validateEmail(email: string) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (email.trim() === "" || !emailRegex.test(email)) {
    return false;
  }
  return true;
}

function validatePhone(phone: string) {
  const phoneRegex = /^[6-9]\d{9}$/;
  if (phone.trim() === "" || !phoneRegex.test(phone)) {
    return false;
  }
  return true;
}

function validatePassword(password: string) {
  if (password.length < 8 || password.trim() === "") {
    return false;
  }
  return true;
}

function validateText(text: string) {
  const nameRegex = /^[A-Za-z\s]+$/;
  if (text.trim() === "" || !nameRegex.test(text)) {
    return false;
  }
  return true;
}

const validateError = (value: string, type: string) => {
  if (value === "") return false;
  switch (type) {
    case "text":
      return !validateText(value);
    case "email":
      return !validateEmail(value);
    case "tel":
      return !validatePhone(value);
    case "password":
    case "confirmPassword":
      return !validatePassword(value);
    default:
      return false;
  }
};
const getHelperText = (
  value: string,
  type: string,
  confirmPassword?: string
): string => {
  if (validateError(value, type)) {
    switch (type) {
      case "text":
        return "Name cannot have special characters";
      case "email":
        return "Enter a valid email";
      case "tel":
        return "Enter a valid phone number";
      case "password":
      case "confirmPassword": {
        return "Password must contain atleast 8 characters";
      }
      default:
        return "Invalid input";
    }
  } else if (type === "confirmPassword" && value && value !== confirmPassword) {
    return "Password does not match";
  } else return "";
};

export {
  validateEmail,
  validatePhone,
  validatePassword,
  validateText,
  validateError,
  getHelperText,
};
