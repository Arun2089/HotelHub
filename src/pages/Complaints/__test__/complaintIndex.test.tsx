import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Complaints from "../index";
import AuthProviderMock from "@src/test-utils/AuthContextMock";
import { fetchData } from "@src/utils";

jest.mock("@src/utils/index.ts", () => ({
  ...jest.requireActual("@src/utils/index.ts"),
  fetchData: jest.fn().mockResolvedValue({
    data: [
      {
        id: 1,
        attributes: {
          type: "Complaint 1",
          date: "2022-01-01",
          description: "This is complaint 1",
          status: "pending",
          studentName: "Test Name 1",
        },
      },
      {
        id: 2,
        attributes: {
          type: "Complaint 2",
          date: "2022-01-02",
          description: "This is complaint 2",
          status: "resolved",
          studentName: "Test Name 2",
        },
      },
      {
        id: 3,
        attributes: {
          type: "Complaint 3",
          date: "2022-01-03",
          description: "This is complaint 3",
          status: "invalid",
          studentName: "Test Name 3",
        },
      },
    ],
  }),
  sendData: jest.fn(),
  deleteData: jest.fn(),
}));

describe("Complaints Page", () => {
  beforeEach(() => {
    render(
      <AuthProviderMock value={{ isLoading: false }}>
        <Complaints />
      </AuthProviderMock>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("snapshot test", async () => {
    const { asFragment } = render(
      <AuthProviderMock value={{ isLoading: false }}>
        <Complaints />
      </AuthProviderMock>
    );
    await waitFor(() => expect(asFragment()).toMatchSnapshot());
  });

  it("renders the complaints page correctly", () => {
    const complaintTable = screen.getByRole("grid");
    expect(fetchData).toHaveBeenCalledTimes(1);
    expect(complaintTable).toBeInTheDocument();
  });

  it("renders the complaints table with correct Headers", async () => {
    expect(screen.getByText("Id")).toBeInTheDocument();
    expect(screen.getByText("Type")).toBeInTheDocument();
    expect(screen.getByText("Created at")).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByText("Complaint 1")).toBeInTheDocument();
      expect(screen.getByText("Complaint 2")).toBeInTheDocument();
      expect(screen.getByText("Complaint 3")).toBeInTheDocument();
    });
  });
});

describe("Complaints Page with no complaints", () => {
  it("renders the complaints table with correct Headers", async () => {
    (fetchData as jest.Mock).mockResolvedValue({
      data: [],
    });
    render(
      <AuthProviderMock value={{ isLoading: false }}>
        <Complaints />
      </AuthProviderMock>
    );

    await waitFor(() => {
      expect(screen.getByText("No rows")).toBeInTheDocument();
    });
  });
});
