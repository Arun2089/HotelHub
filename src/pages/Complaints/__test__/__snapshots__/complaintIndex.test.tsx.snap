// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complaints Page snapshot test 1`] = `
<DocumentFragment>
  <div
    class="flex w-[96%] justify-start items-center gap-2 shrink flex-wrap mx-auto gap-x-2 md:gap-x-4 MuiBox-root css-0"
  >
    <div
      class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault w-24 justify-around capitalize animate-slidein border-primary-light bg-primary-light  bg-opacity-25 border-opacity-25 text-primary-dark flex-row-reverse  css-93yb24-MuiButtonBase-root-MuiChip-root"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
      >
        all
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault w-24 justify-around capitalize animate-slidein border-error-light bg-white text-error-dark flex-row-reverse  css-1izds7p-MuiButtonBase-root-MuiChip-root"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label MuiChip-labelMedium css-1jzq0dw-MuiChip-label"
      >
        pending
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault w-24 justify-around capitalize animate-slidein border-success-light bg-white text-success-dark flex-row-reverse  css-1izds7p-MuiButtonBase-root-MuiChip-root"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label MuiChip-labelMedium css-1jzq0dw-MuiChip-label"
      >
        resolved
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault w-24 justify-around capitalize animate-slidein border-gray-500 bg-white text-gray-600 flex-row-reverse  css-1izds7p-MuiButtonBase-root-MuiChip-root"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label MuiChip-labelMedium css-1jzq0dw-MuiChip-label"
      >
        invalid
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </div>
  <div
    class="w-[96%] h-[85vh] md:h-full my-4 xl:my-6  mx-auto overflow-hidden rounded-xl MuiBox-root css-0"
  >
    <div
      aria-colcount="6"
      aria-multiselectable="false"
      aria-rowcount="1"
      class="max-h-full bg-white rounded-xl MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-13dniha-MuiDataGrid-root"
      role="grid"
    >
      <div
        class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
      >
        <div
          class="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
          role="presentation"
          style="min-height: 56px; max-height: 56px; line-height: 56px;"
        >
          <div
            class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
            role="rowgroup"
            style="transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-rowindex="1"
              class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
              role="row"
            >
              <div
                aria-colindex="1"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor pl-8"
                data-field="id"
                role="columnheader"
                style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                tabindex="0"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                  role="presentation"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 font-semibold css-ahj2mt-MuiTypography-root"
                      >
                        Id
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex="2"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor pl-8"
                data-field="date"
                role="columnheader"
                style="height: 56px; width: 150px; min-width: 150px; max-width: 150px;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                  role="presentation"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 font-semibold css-ahj2mt-MuiTypography-root"
                      >
                        Created at
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex="3"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor"
                data-field="type"
                role="columnheader"
                style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                  role="presentation"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 font-semibold css-ahj2mt-MuiTypography-root"
                      >
                        Type
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDataGrid-virtualScroller css-qvtrhg-MuiDataGrid-virtualScroller"
          role="presentation"
          style="overflow-x: hidden;"
        >
          <div
            class="MuiDataGrid-overlayWrapper css-19qgr0z-MuiDataGrid-overlayWrapper"
          >
            <div
              class="MuiDataGrid-overlayWrapperInner css-1akuw9y-MuiDataGrid-overlayWrapperInner"
              style="height: 0px; width: 0px;"
            >
              <div
                class="MuiDataGrid-overlay css-1w53k9d-MuiDataGrid-overlay"
              >
                <span
                  class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
            role="presentation"
            style="width: auto; height: 1px; min-height: auto;"
          >
            <div
              class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
              role="rowgroup"
              style="transform: translate3d(0px, undefinedpx, 0px);"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-wop1k0-MuiDataGrid-footerContainer"
      >
        <div />
        <div
          class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
          >
            <div
              class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
            />
            <p
              class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
            >
              0–0 of 0
            </p>
            <div
              class="MuiTablePagination-actions"
            >
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to previous page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to next page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
